middleware/capacity_test.go

```diff
@@ -0,0 +1,60 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/julienschmidt/httprouter"
+	"github.com/livepeer/catalyst-api/pipeline"
+	"github.com/stretchr/testify/require"
+)
+
+func TestItCallsNextMiddlewareWhenCapacityAvailable(t *testing.T) {
+	// Create a next handler in the middleware chain, to confirm the request was passed onwards
+	var nextCalled bool
+	next := func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+		nextCalled = true
+	}
+
+	// Set up the HTTP handler
+	handler := HasCapacity(pipeline.NewStubCoordinator(), next)
+
+	// Call the handler
+	responseRecorder := httptest.NewRecorder()
+	handler(responseRecorder, nil, nil)
+
+	// Confirm we got a success reponse and that the handler called the next middleware
+	require.Equal(t, http.StatusOK, responseRecorder.Code)
+	require.True(t, nextCalled)
+}
+
+func TestItErrorsWhenNoCapacityAvailable(t *testing.T) {
+	// Create a next handler in the middleware chain, to confirm the request was passed onwards
+	var nextCalled bool
+	next := func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+		nextCalled = true
+	}
+
+	// Create a lot of in-flight jobs
+	coordinator := pipeline.NewStubCoordinator()
+	for x := 0; x < 5; x++ {
+		coordinator.StartUploadJob(pipeline.UploadJobPayload{
+			RequestID: fmt.Sprintf("request-%d", x),
+		})
+	}
+
+	// Set up the HTTP handler
+	handler := HasCapacity(coordinator, next)
```

Comment:
Hmm I wonder how racy this is since the jobs run in background and the stub handlers return immediately. Does it pass consistently?

Reply:
Yep, I think because the stub handlers never trigger the Mist callbacks right? So the jobs stay in flight

Reply:
Not sure actually. I think they may return an error which finishes the job and sends a callback

Reply:
> it's the real handlers with mock clients instead